File: benchmarks/BunnymarkV1DrawTexture/nim/BunnymarkV1DrawTexture.nim
import random
import godot, engine
import node_2d, texture, sprite, viewport, label, resource_loader

gdobj BunnymarkV1DrawTexture of Node2d:
    var texture: Texture = Texture(load("res://images/godot_bunny.png"))
    var bunnies: seq[tuple[position: Vector2, motion: Vector2]] = @[]
    var gravity: float = 500.0
    var screenSize: Vector2
    
    method ready*() =
        self.setProcess(true)

    method draw*() =
        for bunny in self.bunnies:
            self.drawTexture(self.texture, bunny.position, normalMap=self.texture)

    method process*(delta: float64) =
        self.screenSize = self.getViewportRect().size

        for bunny in self.bunnies.mitems:
            var position = bunny.position
            var motion = bunny.motion

            position.x += motion.x * delta
            position.y += motion.y * delta

            motion.y += self.gravity * delta

            if position.x > self.screenSize.x:
                motion.x *= -1
                position.x = self.screenSize.x
            
            if position.x < 0:
                motion.x *= -1
                position.x = 0
            
            if position.y > self.screenSize.y:
                position.y = self.screenSize.y
                if (random(1.0) > 0.5):
                    motion.y = (random(1100.0) + 50.0)
                else:
                    motion.y *= -0.85
            
            if position.y < 0:
                motion.y = 0
                position.y = 0
            
            bunny.position = position
            bunny.motion = motion
        self.update()

    proc addBunny*() {.gdExport.} =
        self.bunnies.add((vec2(self.screenSize.x / 2, self.screenSize.y / 2), vec2(random(200.0) + 50.0, random(200.0) + 50.0)))
    
    proc removeBunny*() {.gdExport.} =
        if self.bunnies.len == 0:
            return

        var bunny = self.bunnies.pop()
    
    proc finish*() {.gdExport.} =
        self.emitSignal("benchmark_finished", @[newVariant(self.bunnies.len)])

File: benchmarks/BunnymarkV1DrawTexture/es/BunnymarkV1DrawTexture.ts

export default class BunnymarkV1DrawTexture extends godot.Node2D {
	private bunnies: godot.Vector2[][]= [];
	private grav = 500;
	private bunny_texture = godot.ResourceLoader.load("res://images/godot_bunny.png") as godot.Texture;
	private screen_size: godot.Vector2 = null;

	_draw() {
		for (const bunny of this.bunnies) {
			this.draw_texture(this.bunny_texture, bunny[0]);
		}
	}
	
	_process(delta: number) {
		this.screen_size = this.get_viewport_rect().size;
		
		for (let bunny of this.bunnies) {
			var pos = bunny[0];
			var newPosition = bunny[1];
			
			pos.x += newPosition.x * delta;
			pos.y += newPosition.y * delta;
			newPosition.y += this.grav * delta;
		
			if (pos.x > this.screen_size.x) {
				newPosition.x *= -1;
				pos.x = this.screen_size.x;
			}
			
			if (pos.x < 0) {
				newPosition.x *= -1;
				pos.x = 0;
			}
				
			if (pos.y > this.screen_size.y) {
				pos.y = this.screen_size.y;
				if (Math.random() > 0.5)
					newPosition.y = -(godot.randi() % 1100 + 50);
				else
					newPosition.y *= -0.85;
			}
				
			if (pos.y < 0) {
				newPosition.y = 0;
				pos.y = 0;
			}
			
			bunny[0] = pos;
			bunny[1] = newPosition;
		}
		this.update();
	}
	
	add_bunny() {
		this.bunnies.push([
			new godot.Vector2(this.screen_size.x / 2, this.screen_size.y / 2),
			new godot.Vector2(godot.randi() % 200 + 50, godot.randi() % 200 + 50)
		]);
	}
	
	remove_bunny(){
		if (this.bunnies.length == 0)
			return;
		this.bunnies.pop();
	}
	
	finish() {
		this.emit_signal("benchmark_finished", this.bunnies.length);
	}
}
godot.register_class(BunnymarkV1DrawTexture, "BunnymarkV1DrawTexture");